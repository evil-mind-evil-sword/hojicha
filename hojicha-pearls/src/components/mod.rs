//! UI Components for building terminal interfaces
//!
//! This module provides pre-built components similar to Bubbles for Bubbletea,
//! including text input, scrollable views, and more.

#[macro_use]
mod render_guard;
pub use render_guard::SafeRender;

pub mod button;
pub mod help;
pub mod keybinding;
pub mod list;
pub mod modal;
pub mod paginator;
pub mod progress_bar;
pub mod spinner;
pub mod status_bar;
pub mod stopwatch;
pub mod styled_list;
pub mod styled_table;
pub mod table;
pub mod tabs;
pub mod text_input;
pub mod textarea;
pub mod timer;
pub mod utils;
pub mod viewport;

// New unified components that replace duplicates
pub mod unified_list;
pub mod unified_table;

pub use button::{Button, ButtonSize, ButtonVariant};
pub use help::{Help, HelpBuilder, HelpEntry, HelpMode};
pub use keybinding::{KeyBinding, KeyMap};
// Legacy components (deprecated - use unified versions instead)
#[deprecated(note = "Use unified_list::UnifiedList instead")]
pub use list::{List, ListOptions};
#[deprecated(note = "Use unified_table::UnifiedTable instead")]  
pub use table::{Table, TableOptions, TableRow};
#[deprecated(note = "Use unified_list::UnifiedList instead")]
pub use styled_list::{ListItemTrait, StyledList};
#[deprecated(note = "Use unified_table::UnifiedTable instead")]
pub use styled_table::{Column, SortDirection, SortState, StyledTable};

// New unified components (recommended)
pub use unified_list::{UnifiedList, ListItem, ListConfig};
pub use unified_table::{UnifiedTable, TableRow as UnifiedTableRow, Column as UnifiedColumn, SortDirection as UnifiedSortDirection, SortState as UnifiedSortState, TableConfig};

// Other components
pub use modal::{Modal, ModalSize};
pub use paginator::{Paginator, PaginatorStyle};
pub use progress_bar::{ProgressBar, ProgressStyle};
pub use spinner::{Spinner, SpinnerStyle};
pub use status_bar::{StatusBar, StatusBarBuilder, StatusBarPosition, StatusSegment};
pub use stopwatch::{Lap, Stopwatch, StopwatchFormat, StopwatchState};
pub use tabs::{Tab, TabPosition, TabStyle, Tabs, TabsBuilder};
pub use text_input::{TextInput, ValidationResult};
pub use textarea::{TextArea, TextAreaOptions};
pub use timer::{Timer, TimerFormat, TimerState};
pub use viewport::{Viewport, ViewportOptions};
