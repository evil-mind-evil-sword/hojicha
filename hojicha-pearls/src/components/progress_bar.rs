//! Progress bar component with theme support
//!
//! Visual progress indicators with various styles.

use crate::style::{Color, ColorProfile, Style, Theme};
use ratatui::{
    layout::Rect,
    text::{Line, Span},
    widgets::{Gauge, LineGauge, Sparkline},
    Frame,
};

/// Progress bar style variant
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProgressStyle {
    /// Standard filled bar
    Bar,
    /// Line-based progress
    Line,
    /// Sparkline graph
    Spark,
    /// Custom characters for filled/empty
    Custom {
        /// Character to use for filled portion
        filled: char,
        /// Character to use for empty portion
        empty: char,
    },
}

/// Progress bar component
#[derive(Clone)]
pub struct ProgressBar {
    /// Current progress value (0.0 to 1.0)
    value: f64,
    /// Progress history for sparkline
    history: Vec<u64>,
    /// Maximum history length
    max_history: usize,
    /// Progress bar style
    style_variant: ProgressStyle,
    /// Label to display
    label: Option<String>,
    /// Show percentage
    show_percentage: bool,
    /// Show fraction (e.g., "3/10")
    show_fraction: bool,
    /// Total items (for fraction display)
    total: Option<usize>,
    /// Current item (for fraction display)
    current: Option<usize>,
    /// Bar style
    bar_style: Style,
    /// Background style
    background_style: Style,
    /// Label style
    label_style: Style,
    /// Container style
    container_style: Style,
    /// Use gradient colors
    use_gradient: bool,
}

impl ProgressBar {
    /// Create a new progress bar
    pub fn new() -> Self {
        Self {
            value: 0.0,
            history: Vec::new(),
            max_history: 100,
            style_variant: ProgressStyle::Bar,
            label: None,
            show_percentage: true,
            show_fraction: false,
            total: None,
            current: None,
            bar_style: Style::new().fg(Color::green()),
            background_style: Style::new().fg(Color::gray()),
            label_style: Style::new().bold(),
            container_style: Style::new(),
            use_gradient: false,
        }
    }

    /// Set the progress value (0.0 to 1.0)
    pub fn set_progress(&mut self, value: f64) {
        self.value = value.clamp(0.0, 1.0);

        // Add to history for sparkline
        let history_value = (self.value * 100.0) as u64;
        self.history.push(history_value);

        // Trim history if needed
        if self.history.len() > self.max_history {
            self.history.remove(0);
        }
    }

    /// Set progress as fraction
    pub fn set_fraction(&mut self, current: usize, total: usize) {
        self.current = Some(current);
        self.total = Some(total);
        if total > 0 {
            self.value = current as f64 / total as f64;
        } else {
            self.value = 0.0;
        }

        // Update history
        let history_value = (self.value * 100.0) as u64;
        self.history.push(history_value);
        if self.history.len() > self.max_history {
            self.history.remove(0);
        }
    }

    /// Increment progress
    pub fn increment(&mut self, amount: f64) {
        self.set_progress(self.value + amount);
    }

    /// Set the style variant
    pub fn with_style_variant(mut self, variant: ProgressStyle) -> Self {
        self.style_variant = variant;
        self
    }

    /// Set the label
    pub fn with_label(mut self, label: impl Into<String>) -> Self {
        self.label = Some(label.into());
        self
    }

    /// Set whether to show percentage
    pub fn with_percentage(mut self, show: bool) -> Self {
        self.show_percentage = show;
        self
    }

    /// Set whether to show fraction
    pub fn with_fraction(mut self, show: bool) -> Self {
        self.show_fraction = show;
        self
    }

    /// Set bar style
    pub fn with_bar_style(mut self, style: Style) -> Self {
        self.bar_style = style;
        self
    }

    /// Set background style
    pub fn with_background_style(mut self, style: Style) -> Self {
        self.background_style = style;
        self
    }

    /// Set container style
    pub fn with_container_style(mut self, style: Style) -> Self {
        self.container_style = style;
        self
    }

    /// Enable gradient colors
    pub fn with_gradient(mut self, enable: bool) -> Self {
        self.use_gradient = enable;
        self
    }

    /// Set maximum history length for sparkline
    pub fn with_max_history(mut self, max: usize) -> Self {
        self.max_history = max;
        self
    }

    /// Get the current progress value
    pub fn value(&self) -> f64 {
        self.value
    }

    /// Check if progress is complete
    pub fn is_complete(&self) -> bool {
        self.value >= 1.0
    }

    /// Reset the progress bar
    pub fn reset(&mut self) {
        self.value = 0.0;
        self.current = None;
        self.history.clear();
    }

    /// Get dynamic color based on progress
    fn get_progress_color(&self, theme: &Theme) -> Color {
        if !self.use_gradient {
            return self
                .bar_style
                .get_foreground()
                .cloned()
                .unwrap_or_else(|| theme.colors.primary.clone());
        }

        // Gradient from red -> yellow -> green based on progress
        if self.value < 0.33 {
            theme.colors.error.clone()
        } else if self.value < 0.66 {
            theme.colors.warning.clone()
        } else {
            theme.colors.success.clone()
        }
    }

    /// Build the label text
    fn build_label(&self) -> String {
        let mut parts = Vec::new();

        if let Some(ref label) = self.label {
            parts.push(label.clone());
        }

        if self.show_fraction && self.current.is_some() && self.total.is_some() {
            parts.push(format!("{}/{}", self.current.unwrap(), self.total.unwrap()));
        }

        if self.show_percentage {
            parts.push(format!("{:.0}%", self.value * 100.0));
        }

        parts.join(" ")
    }

    /// Apply a theme to this progress bar
    pub fn apply_theme(&mut self, theme: &Theme) {
        self.bar_style = Style::new().fg(theme.colors.primary.clone());

        self.background_style = Style::new().fg(theme.colors.surface.clone());

        self.label_style = Style::new().fg(theme.colors.text.clone()).bold();

        if self.use_gradient {
            // Gradient will be applied dynamically based on value
        }
    }

    /// Render the progress bar
    pub fn render(&self, frame: &mut Frame, area: Rect, theme: &Theme, profile: &ColorProfile) {
        if !super::utils::is_valid_area(area) {
            return;
        }

        let label_text = self.build_label();
        let progress_color = self.get_progress_color(theme);

        match self.style_variant {
            ProgressStyle::Bar => {
                let gauge = Gauge::default()
                    .label(label_text)
                    .percent((self.value * 100.0) as u16)
                    .gauge_style(Style::new().fg(progress_color).to_ratatui(profile))
                    .style(self.background_style.to_ratatui(profile));

                frame.render_widget(gauge, area);
            }

            ProgressStyle::Line => {
                let line_gauge = LineGauge::default()
                    .label(label_text)
                    .ratio(self.value)
                    .line_set(ratatui::symbols::line::THICK)
                    .filled_style(Style::new().fg(progress_color).to_ratatui(profile))
                    .style(self.background_style.to_ratatui(profile));

                frame.render_widget(line_gauge, area);
            }

            ProgressStyle::Spark => {
                // Use sparkline for history visualization
                // Ensure we have some data to display
                let data = if self.history.is_empty() {
                    vec![0]
                } else {
                    self.history.clone()
                };

                let sparkline = Sparkline::default()
                    .data(&data)
                    .style(Style::new().fg(progress_color).to_ratatui(profile));

                // If we have a label, split the area
                if !label_text.is_empty() {
                    use ratatui::layout::{Constraint, Direction, Layout};

                    let chunks = Layout::default()
                        .direction(Direction::Vertical)
                        .constraints([Constraint::Length(1), Constraint::Min(1)])
                        .split(area);

                    // Render label
                    let label_line = Line::from(Span::styled(
                        label_text,
                        self.label_style.to_ratatui(profile),
                    ));
                    frame.render_widget(ratatui::widgets::Paragraph::new(label_line), chunks[0]);

                    // Render sparkline
                    frame.render_widget(sparkline, chunks[1]);
                } else {
                    frame.render_widget(sparkline, area);
                }
            }

            ProgressStyle::Custom { filled, empty } => {
                // Custom character-based progress bar
                let bar_width = area.width.saturating_sub(2) as usize; // Account for borders/padding
                if bar_width == 0 {
                    return;
                }
                let filled_count = (bar_width as f64 * self.value).min(bar_width as f64) as usize;
                let empty_count = bar_width.saturating_sub(filled_count);

                let filled_part = filled.to_string().repeat(filled_count);
                let empty_part = empty.to_string().repeat(empty_count);

                let content = if !label_text.is_empty() {
                    vec![
                        Line::from(Span::styled(
                            label_text,
                            self.label_style.to_ratatui(profile),
                        )),
                        Line::from(vec![
                            Span::styled(
                                filled_part.clone(),
                                Style::new().fg(progress_color.clone()).to_ratatui(profile),
                            ),
                            Span::styled(
                                empty_part.clone(),
                                self.background_style.to_ratatui(profile),
                            ),
                        ]),
                    ]
                } else {
                    vec![Line::from(vec![
                        Span::styled(
                            filled_part,
                            Style::new().fg(progress_color.clone()).to_ratatui(profile),
                        ),
                        Span::styled(empty_part, self.background_style.to_ratatui(profile)),
                    ])]
                };

                let paragraph = ratatui::widgets::Paragraph::new(content);
                frame.render_widget(paragraph, area);
            }
        }
    }
}

impl Default for ProgressBar {
    fn default() -> Self {
        Self::new()
    }
}
