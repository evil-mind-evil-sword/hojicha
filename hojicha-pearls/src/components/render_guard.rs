//! Render guard macro for safe component rendering
//!
//! Provides a macro to ensure components don't crash when areas have zero dimensions.

/// Macro to guard render functions against invalid areas
///
/// Usage:
/// ```
/// render_guard!(area);
/// ```
///
/// This will return early from the function if the area has zero width or height.
#[macro_export]
macro_rules! render_guard {
    ($area:expr) => {
        if $area.width == 0 || $area.height == 0 {
            return;
        }
    };

    // Version that returns a value
    ($area:expr, $return_value:expr) => {
        if $area.width == 0 || $area.height == 0 {
            return $return_value;
        }
    };
}

/// Trait for safe rendering with automatic area validation
pub trait SafeRender {
    /// Render with automatic area validation
    fn safe_render<F, R>(&self, area: ratatui::layout::Rect, render_fn: F) -> Option<R>
    where
        F: FnOnce() -> R,
    {
        if area.width > 0 && area.height > 0 {
            Some(render_fn())
        } else {
            None
        }
    }
}

// Implement for all types
impl<T> SafeRender for T {}
