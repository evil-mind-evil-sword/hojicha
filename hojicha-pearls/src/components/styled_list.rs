//! Styled list component with theme support
//!
//! **DEPRECATED**: This component has been superseded by `unified_list::UnifiedList`.
//! Use `UnifiedList` for new code as it provides all the functionality of both
//! `List` and `StyledList` in a single, configurable component.
//!
//! A scrollable list with selection, filtering, and rich styling options.

use crate::style::{ColorProfile, Style, Theme};
use hojicha_core::event::{Event, Key, KeyEvent};
use ratatui::{
    layout::Rect,
    style::{Modifier, Style as RatatuiStyle},
    text::{Line, Span},
    widgets::{Block, Borders, List as RatatuiList, ListItem as RatatuiListItem, ListState},
    Frame,
};

/// List item that can be displayed
pub trait ListItemTrait: Clone {
    /// Get the display text for this item
    fn display(&self) -> String;

    /// Get the search text for filtering (defaults to display text)
    fn search_text(&self) -> String {
        self.display()
    }
}

impl ListItemTrait for String {
    fn display(&self) -> String {
        self.clone()
    }
}

impl ListItemTrait for &str {
    fn display(&self) -> String {
        self.to_string()
    }
}

impl ListItemTrait for i32 {
    fn display(&self) -> String {
        self.to_string()
    }
}

impl ListItemTrait for usize {
    fn display(&self) -> String {
        self.to_string()
    }
}

/// A styled, scrollable list component
pub struct StyledList<T: ListItemTrait> {
    /// The items in the list
    items: Vec<T>,
    /// Filtered items (subset of items based on filter)
    filtered_items: Vec<T>,
    /// Current filter string
    filter: String,
    /// Whether filtering is enabled
    filter_enabled: bool,
    /// The list state (selection, scroll position)
    state: ListState,
    /// Title of the list
    title: Option<String>,
    /// Style for normal items
    item_style: Style,
    /// Style for the selected item
    selected_style: Style,
    /// Style for the list container
    container_style: Style,
    /// Style for the title
    title_style: Style,
    /// Style for the filter indicator
    filter_style: Style,
    /// Whether the list is focused
    focused: bool,
    /// Show selection indicator
    show_selection: bool,
    /// Selection indicator character
    selection_indicator: String,
    /// Maximum height (in rows)
    max_height: Option<u16>,
}

impl<T: ListItemTrait> StyledList<T> {
    /// Create a new styled list
    pub fn new(items: Vec<T>) -> Self {
        let mut state = ListState::default();
        if !items.is_empty() {
            state.select(Some(0));
        }

        let filtered_items = items.clone();

        Self {
            items: items.clone(),
            filtered_items,
            filter: String::new(),
            filter_enabled: false,
            state,
            title: None,
            item_style: Style::new(),
            selected_style: Style::new().bg(crate::style::Color::blue()).bold(),
            container_style: Style::new().border(crate::style::BorderStyle::Normal),
            title_style: Style::new().bold(),
            filter_style: Style::new().fg(crate::style::Color::yellow()).italic(),
            focused: false,
            show_selection: true,
            selection_indicator: "> ".to_string(),
            max_height: None,
        }
    }

    /// Set the title
    pub fn with_title(mut self, title: impl Into<String>) -> Self {
        self.title = Some(title.into());
        self
    }

    /// Enable or disable filtering
    pub fn with_filter(mut self, enabled: bool) -> Self {
        self.filter_enabled = enabled;
        self
    }

    /// Set the filter string
    pub fn set_filter(&mut self, filter: String) {
        self.filter = filter.to_lowercase();
        self.apply_filter();
    }

    /// Clear the filter
    pub fn clear_filter(&mut self) {
        self.filter.clear();
        self.apply_filter();
    }

    /// Apply the current filter to items
    fn apply_filter(&mut self) {
        if self.filter.is_empty() {
            self.filtered_items = self.items.clone();
        } else {
            self.filtered_items = self
                .items
                .iter()
                .filter(|item| item.search_text().to_lowercase().contains(&self.filter))
                .cloned()
                .collect();
        }

        // Reset selection if needed
        if self.filtered_items.is_empty() {
            self.state.select(None);
        } else if self
            .state
            .selected()
            .map_or(true, |i| i >= self.filtered_items.len())
        {
            self.state.select(Some(0));
        }
    }

    /// Update the items in the list
    pub fn set_items(&mut self, items: Vec<T>) {
        self.items = items;
        self.apply_filter();
    }

    /// Get the currently selected item
    pub fn selected(&self) -> Option<&T> {
        self.state
            .selected()
            .and_then(|i| self.filtered_items.get(i))
    }

    /// Get the index of the selected item
    pub fn selected_index(&self) -> Option<usize> {
        self.state.selected()
    }

    /// Select a specific item by index
    pub fn select(&mut self, index: Option<usize>) {
        if let Some(i) = index {
            if i < self.filtered_items.len() {
                self.state.select(Some(i));
            }
        } else {
            self.state.select(None);
        }
    }

    /// Move selection up
    pub fn select_previous(&mut self) {
        let current = self.state.selected().unwrap_or(0);
        let new_index = if current == 0 {
            self.filtered_items.len().saturating_sub(1)
        } else {
            current.saturating_sub(1)
        };

        if !self.filtered_items.is_empty() {
            self.state.select(Some(new_index));
        }
    }

    /// Move selection down
    pub fn select_next(&mut self) {
        let current = self.state.selected().unwrap_or(0);
        let new_index = if current >= self.filtered_items.len().saturating_sub(1) {
            0
        } else {
            current + 1
        };

        if !self.filtered_items.is_empty() {
            self.state.select(Some(new_index));
        }
    }

    /// Move to the first item
    pub fn select_first(&mut self) {
        if !self.filtered_items.is_empty() {
            self.state.select(Some(0));
        }
    }

    /// Move to the last item
    pub fn select_last(&mut self) {
        if !self.filtered_items.is_empty() {
            self.state.select(Some(self.filtered_items.len() - 1));
        }
    }

    /// Set focused state
    pub fn focus(&mut self) {
        self.focused = true;
    }

    /// Remove focus
    pub fn blur(&mut self) {
        self.focused = false;
    }

    /// Check if focused
    pub fn is_focused(&self) -> bool {
        self.focused
    }

    /// Set the item style
    pub fn with_item_style(mut self, style: Style) -> Self {
        self.item_style = style;
        self
    }

    /// Set the selected item style
    pub fn with_selected_style(mut self, style: Style) -> Self {
        self.selected_style = style;
        self
    }

    /// Set the container style
    pub fn with_container_style(mut self, style: Style) -> Self {
        self.container_style = style;
        self
    }

    /// Set the title style
    pub fn with_title_style(mut self, style: Style) -> Self {
        self.title_style = style;
        self
    }

    /// Set whether to show selection indicator
    pub fn with_selection_indicator(mut self, show: bool) -> Self {
        self.show_selection = show;
        self
    }

    /// Set custom selection indicator
    pub fn with_custom_indicator(mut self, indicator: impl Into<String>) -> Self {
        self.selection_indicator = indicator.into();
        self
    }

    /// Set maximum height
    pub fn with_max_height(mut self, height: u16) -> Self {
        self.max_height = Some(height);
        self
    }

    /// Handle keyboard input
    pub fn handle_event(&mut self, event: Event<()>) -> bool {
        if !self.focused {
            return false;
        }

        match event {
            Event::Key(KeyEvent { key, .. }) => match key {
                Key::Up | Key::Char('k') => {
                    self.select_previous();
                    true
                }
                Key::Down | Key::Char('j') => {
                    self.select_next();
                    true
                }
                Key::Home | Key::Char('g') => {
                    self.select_first();
                    true
                }
                Key::End | Key::Char('G') => {
                    self.select_last();
                    true
                }
                _ => false,
            },
            _ => false,
        }
    }

    /// Apply a theme to this list
    pub fn apply_theme(&mut self, theme: &Theme) {
        if let Some(style) = theme.get_style("list.item") {
            self.item_style = style.clone();
        }

        if let Some(style) = theme.get_style("list.selected") {
            self.selected_style = style.clone();
        }

        if let Some(style) = theme.get_style("list.container") {
            self.container_style = style.clone();
        }

        if let Some(style) = theme.get_style("list.title") {
            self.title_style = style.clone();
        }
    }

    /// Render the list
    pub fn render(&mut self, frame: &mut Frame, area: Rect, profile: &ColorProfile) {
        // Create list items with styling
        let items: Vec<RatatuiListItem> = self
            .filtered_items
            .iter()
            .enumerate()
            .map(|(i, item)| {
                let is_selected = self.state.selected() == Some(i);
                let mut text = item.display();

                if is_selected && self.show_selection {
                    text = format!("{}{}", self.selection_indicator, text);
                }

                let style = if is_selected {
                    self.selected_style.to_ratatui(profile)
                } else {
                    self.item_style.to_ratatui(profile)
                };

                RatatuiListItem::new(Line::from(Span::styled(text, style)))
            })
            .collect();

        // Create the block with title and filter indicator
        let mut block = Block::default();

        if self.container_style.get_border() != &crate::style::BorderStyle::None {
            block = block
                .borders(Borders::ALL)
                .border_type(self.container_style.get_border().to_ratatui());

            if let Some(border_color) = self.container_style.get_border_color() {
                let mut border_style = RatatuiStyle::default().fg(border_color.to_ratatui(profile));

                if self.focused {
                    border_style = border_style.add_modifier(Modifier::BOLD);
                }

                block = block.border_style(border_style);
            }
        }

        // Add title with filter indicator
        if let Some(ref title) = self.title {
            let title_text = if self.filter_enabled && !self.filter.is_empty() {
                format!("{} [filter: {}]", title, self.filter)
            } else {
                title.clone()
            };

            block = block
                .title(title_text)
                .title_style(self.title_style.to_ratatui(profile));
        } else if self.filter_enabled && !self.filter.is_empty() {
            block = block
                .title(format!("[filter: {}]", self.filter))
                .title_style(self.filter_style.to_ratatui(profile));
        }

        // Create the list widget
        let list = RatatuiList::new(items)
            .block(block)
            .style(self.container_style.to_ratatui(profile));

        // Apply max height if set
        let render_area = if let Some(max_h) = self.max_height {
            Rect {
                x: area.x,
                y: area.y,
                width: area.width,
                height: area.height.min(max_h),
            }
        } else {
            area
        };

        // Render the list
        frame.render_stateful_widget(list, render_area, &mut self.state);
    }

    /// Get the number of items (after filtering)
    pub fn len(&self) -> usize {
        self.filtered_items.len()
    }

    /// Check if the list is empty (after filtering)
    pub fn is_empty(&self) -> bool {
        self.filtered_items.is_empty()
    }
}
