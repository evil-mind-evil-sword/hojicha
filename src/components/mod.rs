//! UI Components for building terminal interfaces
//!
//! This module provides pre-built components similar to Bubbles for Bubbletea,
//! including text input, scrollable views, and more.

pub mod keybinding;
pub mod list;
pub mod spinner;
pub mod table;
pub mod textarea;
pub mod viewport;

pub use keybinding::{KeyBinding, KeyMap};
pub use list::{List, ListOptions};
pub use spinner::{Spinner, SpinnerStyle};
pub use table::{Table, TableOptions, TableRow};
pub use textarea::{TextArea, TextAreaOptions};
pub use viewport::{Viewport, ViewportOptions};
