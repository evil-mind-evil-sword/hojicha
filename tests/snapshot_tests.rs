//! Snapshot tests for UI components using insta and ratatui's TestBackend

use hojicha::components::{List, Spinner, SpinnerStyle, Table, TextArea, Viewport};
use insta::assert_snapshot;
use ratatui::{
    Terminal,
    backend::TestBackend,
    buffer::Buffer,
    widgets::{Block, Borders},
};

/// Helper to render a component and capture its buffer as a string
fn render_component<F>(width: u16, height: u16, render_fn: F) -> String
where
    F: FnOnce(&mut ratatui::Frame),
{
    let backend = TestBackend::new(width, height);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(render_fn).unwrap();

    // Get the buffer and convert to string
    let buffer = terminal.backend().buffer();
    buffer_to_string(buffer)
}

/// Convert a buffer to string for snapshot testing
fn buffer_to_string(buffer: &Buffer) -> String {
    let mut result = String::new();
    for y in 0..buffer.area.height {
        for x in 0..buffer.area.width {
            let cell = buffer.get(x, y);
            result.push_str(cell.symbol());
        }
        if y < buffer.area.height - 1 {
            result.push('\n');
        }
    }
    result
}

/// Helper to copy buffer contents
fn copy_buffer_to_frame(src: &Buffer, frame: &mut ratatui::Frame, offset_x: u16, offset_y: u16) {
    // Buffer content is stored as a flat array, indexed from the buffer's area
    let src_area = src.area;
    for y in 0..src_area.height {
        for x in 0..src_area.width {
            // src.get() expects absolute coordinates
            let cell = src.get(src_area.x + x, src_area.y + y);
            let dst_x = offset_x + x;
            let dst_y = offset_y + y;
            if dst_x < frame.size().width && dst_y < frame.size().height {
                let dst_idx = (dst_y * frame.size().width + dst_x) as usize;
                frame.buffer_mut().content[dst_idx] = cell.clone();
            }
        }
    }
}

#[test]
fn test_list_snapshot() {
    let output = render_component(30, 10, |f| {
        let items = vec!["First item", "Second item", "Third item", "Fourth item"];
        let mut list = List::new(items);
        list.select(1); // Select second item

        let area = f.size();
        let block = Block::default().title("List Widget").borders(Borders::ALL);

        // Calculate inner area for the list
        let inner = block.inner(area);
        f.render_widget(block, area);

        // Render list in a buffer first
        let mut list_buffer = Buffer::empty(inner);
        list.render(inner, &mut list_buffer);

        // Copy list buffer to frame
        copy_buffer_to_frame(&list_buffer, f, inner.x, inner.y);
    });

    assert_snapshot!(output);
}

#[test]
fn test_spinner_snapshot() {
    let output = render_component(20, 5, |f| {
        let mut spinner = Spinner::new();
        spinner.set_style(SpinnerStyle::Dots);
        spinner.set_message("Loading...");

        // Tick a few times to get a specific frame
        for _ in 0..3 {
            spinner.tick();
        }

        let area = f.size();
        let mut buffer = Buffer::empty(area);
        spinner.render(area, &mut buffer);

        // Copy to frame
        copy_buffer_to_frame(&buffer, f, 0, 0);
    });

    assert_snapshot!(output);
}

#[test]
fn test_table_snapshot() {
    // Define a simple row type that implements TableRow
    #[derive(Debug, Clone)]
    struct Row {
        id: String,
        name: String,
        status: String,
    }

    impl hojicha::components::table::TableRow for Row {
        fn to_row(&self) -> Vec<String> {
            vec![self.id.clone(), self.name.clone(), self.status.clone()]
        }
    }

    let output = render_component(40, 10, |f| {
        let headers = vec!["ID".to_string(), "Name".to_string(), "Status".to_string()];
        let rows = vec![
            Row {
                id: "1".to_string(),
                name: "Alice".to_string(),
                status: "Active".to_string(),
            },
            Row {
                id: "2".to_string(),
                name: "Bob".to_string(),
                status: "Inactive".to_string(),
            },
            Row {
                id: "3".to_string(),
                name: "Charlie".to_string(),
                status: "Active".to_string(),
            },
        ];

        let mut table = Table::new(headers).with_rows(rows);
        table.select(1); // Select Bob's row

        let area = f.size();
        let block = Block::default().title("Table Widget").borders(Borders::ALL);

        let inner = block.inner(area);
        f.render_widget(block, area);

        let mut table_buffer = Buffer::empty(inner);
        table.render(inner, &mut table_buffer);

        // Copy to frame
        copy_buffer_to_frame(&table_buffer, f, inner.x, inner.y);
    });

    assert_snapshot!(output);
}

#[test]
fn test_textarea_snapshot() {
    let output = render_component(30, 8, |f| {
        let mut textarea = TextArea::new();
        textarea.set_value("Hello, World!\nThis is line 2\nAnd line 3.");

        let area = f.size();
        let block = Block::default().title("TextArea").borders(Borders::ALL);

        let inner = block.inner(area);
        f.render_widget(block, area);

        let mut textarea_buffer = Buffer::empty(inner);
        textarea.render(inner, &mut textarea_buffer);

        // Copy to frame
        copy_buffer_to_frame(&textarea_buffer, f, inner.x, inner.y);
    });

    assert_snapshot!(output);
}

#[test]
fn test_viewport_snapshot() {
    let output = render_component(25, 10, |f| {
        let content = (0..15)
            .map(|i| format!("Line {i}"))
            .collect::<Vec<_>>()
            .join("\n");

        let mut viewport = Viewport::new();
        viewport.set_content(content);

        // Scroll down a bit
        viewport.scroll_down(3);

        let area = f.size();
        let block = Block::default().title("Viewport").borders(Borders::ALL);

        let inner = block.inner(area);
        f.render_widget(block, area);

        let mut viewport_buffer = Buffer::empty(inner);
        viewport.render(inner, &mut viewport_buffer);

        // Copy to frame
        copy_buffer_to_frame(&viewport_buffer, f, inner.x, inner.y);
    });

    assert_snapshot!(output);
}

#[test]
fn test_list_with_selection_snapshot() {
    // Test different selection states
    for selected in [None, Some(0), Some(2)] {
        let output = render_component(20, 5, |f| {
            let items = vec!["Item A", "Item B", "Item C"];
            let mut list = List::new(items);
            if let Some(idx) = selected {
                list.select(idx);
            }

            let area = f.size();
            let mut buffer = Buffer::empty(area);
            list.render(area, &mut buffer);

            // Copy to frame
            copy_buffer_to_frame(&buffer, f, 0, 0);
        });

        let snapshot_name = match selected {
            None => "list_no_selection",
            Some(0) => "list_first_selected",
            Some(2) => "list_third_selected",
            _ => "list_other",
        };

        assert_snapshot!(snapshot_name, output);
    }
}

#[test]
fn test_spinner_styles_snapshot() {
    let styles = [
        (SpinnerStyle::Dots, "dots"),
        (SpinnerStyle::Line, "line"),
        (SpinnerStyle::Circle, "circle"),
        (SpinnerStyle::Square, "square"),
        (SpinnerStyle::Arrow, "arrow"),
    ];

    for (style, name) in styles {
        let output = render_component(15, 3, |f| {
            let mut spinner = Spinner::new();
            spinner.set_style(style);

            // Tick to get a frame
            spinner.tick();

            let area = f.size();
            let mut buffer = Buffer::empty(area);
            spinner.render_centered(area, &mut buffer);

            // Copy to frame
            copy_buffer_to_frame(&buffer, f, 0, 0);
        });

        assert_snapshot!(format!("spinner_{}", name), output);
    }
}
