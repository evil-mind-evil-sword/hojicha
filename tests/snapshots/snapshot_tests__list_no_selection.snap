---
source: tests/snapshot_tests.rs
expression: output
---
Item A              
Item B              
Item C
